core/src/execution/datafusion/expressions/cast.rs

```diff
@@ -1208,6 +1277,260 @@ fn do_cast_string_to_int<
+    let base = T::Native::usize_as(10);
+
+    let bs = s.as_bytes();
+    let (bs, negative) = match bs.first() {
+        Some(b'-') => (&bs[1..], true),
+        Some(b'+') => (&bs[1..], false),
+        _ => (bs, false),
+    };
+
+    if bs.is_empty() {
+        return None;
+    }
+
+    let mut bs = bs.iter();
+    while let Some(b) = bs.next() {
+        match b {
+            b'0'..=b'9' => {
+                if digits == 0 && *b == b'0' {
+                    leading_zeros += 1;
+                    continue;
+                }
+                digits += 1;
+                result = result.mul_wrapping(base);
+                result = result.add_wrapping(T::Native::usize_as((b - b'0') as usize));
+            }
+            b'.' => {
+                while let Some(b) = bs.next() {
+                    if *b == b'e' || *b == b'E' {
+                        match parse_exponent(
+                            bs.by_ref(),
+                            &mut digits,
+                            &mut fractionals,
+                            &mut leading_zeros,
+                            &mut has_exponent,
+                            &mut has_negative_exp,
+                        ) {
+                            None => {
+                                return None;
+                            }
+                            Some(v) => {
+                                exponent = v;
+                                continue;
+                            }
+                        }
+                    }
+                    if !b.is_ascii_digit() {
+                        return None;
+                    }
+                    fractionals += 1;
+                    digits += 1;
+                    result = result.mul_wrapping(base);
+                    result = result.add_wrapping(T::Native::usize_as((b - b'0') as usize));
+                }
+
+                // +00. is a valid decimal value in Spark but +. is invalid.
+                if digits == 0 && leading_zeros == 0 {
+                    return None;
+                }
+            }
+            b'e' | b'E' => match parse_exponent(
+                &mut bs,
+                &mut digits,
+                &mut fractionals,
+                &mut leading_zeros,
+                &mut has_exponent,
+                &mut has_negative_exp,
+            ) {
+                None => {
+                    return None;
+                }
+                Some(v) => {
+                    exponent = v;
+                    continue;
+                }
+            },
+            _ => {
+                return None;
+            }
+        }
+    }
+
+    result = adjust_decimal_scale::<T>(
+        result,
+        precision,
+        scale,
+        base,
+        exponent,
+        fractionals,
+        digits,
+        has_exponent,
+        has_negative_exp,
+    )?;
+
+    return Some(if negative {
+        result.neg_wrapping()
+    } else {
+        result
+    });
```

Comment:
Because this is the last expression, the `return` here is redundant.

```suggestion
    Some(if negative {
        result.neg_wrapping()
    } else {
        result
    })
```